CenterForOpenScience/osf.io
# [ENG-164][ENG-166][ENG-174] DraftNode and DraftRegistration API Endpoints

## Purpose

A combination of three other pull requests to allow for a more efficient code reviewing process. 
* Adds DraftRegistrationList and DraftRegistrationDetail API endpoints to create, retrieve, update, and delete DraftRegistrations
* Updates RegistrationList to allow for requests to create Registrations
* Adds DraftRegistrationRelationship API endpoints to manage DraftRegistration relationships with contributors, institutions, and subjects.

## Changes



### Endpoints
* Registrations - `/registrations/`
  * POST - Creates a registration for the specified draft_registration and nodes
* DraftRegistrationList - `/draft_registrations/`
  * GET - Retrieves all draft_registrations associated with the user
  * POST - Creates draft_registration 
* DraftRegistrationDetail - `/draft_registrations/<draft_id>`
  * GET - Retrieves the details of the specified draft_registration
  * PATCH - Updates details of the specified draft_registration
  * DELETE - Deletes the specified draft_registration
* DraftRegistrationContributorsList - `/draft_registration/<draft_id>/contributors/`
  * GET - Retrieves the contributors on the draft_registration
  * POST - Add a user as a contributor on the draft_registration
* DraftRegistrationContributorsDetail - `/draft_registration/<draft_id>/contributors/<contributor_id>/`
  * GET - Retrieves details about the specified contributor on a specified draft_registration
  * PATCH - Updates details (permissions, bibliographic index) about the specified contributor on a specified draft_registration
  * DELETE - Deletes the contributor relationship (the user continues to exist) between the specified contributor and draft_registration
* DraftInstitutionsList - `/draft_registrations/<draft_id>/institutions/`
  * GET - Retrieves a list of all institutions affiliated with the specified draft_registration
* DraftInstitutionsRelationship - `/draft_registrations/<draft_id>/relationships/institutions/`
  * PUT/PATCH - Updates the set of institutions associated with the specified draft_registration; can be used to dissociate all institutions with the draft_registration by providing an empty list
* DraftSubjectsList - `/draft_registrations/<draft_id>/subjects/`
  * GET - Retrieves a list of all subjects affiliated with the specified draft_registration
* DraftSubjectsRelationship - `/draft_registrations/<draft_id>/relationships/subjects`
  * PUT/PATCH - Updates the set of subjects associated with the specified draft_registration; can be used to dissociate all subjects with the draft_registration by providing an empty list


## QA Notes

This PR does not require a data migration. It builds upon the model changes and migration changes in [this PR](https://github.com/CenterForOpenScience/osf.io/pull/9229) and is generally an additive change. Changes can be verified through the API. Some details about the API endpoints can be found in this [Notion document](https://www.notion.so/cos/Draft-Node-Endpoints-2aefc98b39844e42805e511937a32ff5).

Things to keep an eye out for:
* Existing and currently-used API functionality around the creation, retrieval, and manipulation of draft registrations and registrations (can likely be verified by working through the draft registration and registration creation process on OSF) should continue to behave the same way
* New API Endpoints

## Documentation


developer.osf.io needs to be updated to reflect the changes to and creation of these API endpoints

## Side Effects



## Ticket


* [ENG-164 - DraftRegistrationList and Detail](https://openscience.atlassian.net/browse/ENG-164)
* [ENG-174 - Create Registration](https://openscience.atlassian.net/browse/ENG-174)
* [ENG-166 - DraftRegistrationRelationships](https://openscience.atlassian.net/browse/ENG-166)
